068-access-policies-in-jwt.md

```diff
@@ -0,0 +1,251 @@
+There is currently no inbuilt support for JWT-based authorization.
+If a user wants to use JWT-based authorization, they have to implement their own authorizer and use the custom authorizer feature.
+As the documentation states:
+
+> In addition to the Kafka custom resource configuration, the JAR file containing the custom authorizer class along with its dependencies must be available on the classpath of the Kafka broker.
+
+This proposal aims to add the JWT-based authorizer to Strimzi, so it can be selected in the configuration as
+another option apart from the simple and Keycloak ones, without having to implement a custom authorizer.
+
+## Motivation
+
+Having a generic solution is always preferable as it allows for less effort to be maintained.
+While we cannot achieve a perfect generic solution due to the nature of OAuth2, and how different
+identity providers are implemented, we can strive to create a solution as generic as possible.
+
+Passing additional data in the JWT seems to be supported by the majority of identity providers.
+Therefore, we can use this to pass the access policies in the JWT.
+This solution will work in most cases and with most identity providers, however it will not
+scale to infinity.
+It will, however, be decoupled from Keycloak and therefore be more generic.
+
+## Why JWT as an authorization mechanism?
+
+The official [JWT](https://jwt.io/introduction) lists some use cases:
+
+> Authorization: This is the most common scenario for using JWT. Once the user is logged in, each subsequent request will include the JWT, allowing the user to access routes, services, and resources that are permitted with that token. Single Sign On is a feature that widely uses JWT nowadays, because of its small overhead and its ability to be easily used across different domains.
+
+JWT supports custom claims, which can be used to pass additional claims/roles.
+For example [Spring Security](https://docs.spring.io/spring-authorization-server/reference/guides/how-to-custom-claims-authorities.html) can pass both claims and roles in the JWT,
+which can be used to authorize the user.
+
+All OAuth2 providers support custom claims in the JWT in one way or another.
+We have listed the most popular ones in the [Support for roles or claims in popular identity providers](#support-for-roles-or-claims-in-popular-identity-providers) section with links to the documentation
+on how custom claims or roles are supported.
+
+The concept of roles can be directly mapped to the concept of access policies in Kafka.
+
+## Proposal
+
+Identity providers support passing custom fields in the JWT token.
+Some providers allow the data type to be string, while others allow a list of strings.
+There are not many (if any) identity providers that allow a list of objects.
+
+Therefore, in this proposal we allow the ACLs to be passed as either.
+e.g. for a list of strings:
+```json
+{
+  "acls": [
+    "cluster_x:topic_1:read",
+    ":topic2:write"
```

Comment:
is this syntax a kind of standard? My understanding is that omitting the cluster name, could mean all clusters like using a regex like `.*`. Is it right? Maybe it's related to the @fvaleri question if we are going to support regex and what's the syntax for this acls.

Reply:
It should probably be something like: `CLUSTER_NAME:RESOURCE_TYPE:RESOURCE_SPEC:PERMITTED_ACTIONS`

Specifically:
* `my-cluster:Topic:topic_1:Read,Describe,DescribeConfigs`
ACL specifically for Kafka cluster named `my-cluster`, grants Read, Describe and DescribeConfigs permissions to topic called `topic_1`

* `:Group:prod_consumers_*:Describe,Delete,Read`
ACL grants Describe,Delete,Read to consumer groups that start with `prod_consumers_` on any Kafka cluster

Reply:
I've updated the proposal with more details and examples. This should cover most of the other questions on the ACL syntax in previous threads as well.

I'm using mostly the same format as @mstruk proposes.
I have opted to have the `RESOURCE_TYPE` and `PERMITTED_ACTIONS` be all lowercase.
I'd also change the separator for the `PERMITTED_ACTIONS` to `|` and keep the `,` for separating different ACLs when they are passed in a single string.
However, I have no strong opinion about the exact format.